#Exercise 2: Binning 'YearBuilt' variable from AMES Housing dataset
In this exercise, we will learn how to create a new feature by binning an existing numerical column.

The dataset we will be using in this exercise is the Ames Housing dataset and it can be found on our GitHub repository - https://github.com/PacktWorkshops/The-Data-Science-Workshop/blob/master/Chapter12/Dataset/ames_iowa_housing.csv

Note
The dataset used for this exercise is the Ames Housing dataset compiled by Dean De Cock: https://www.openintro.org/stat/data/?data=ames 

This dataset contains the list of residential home sales in the city of Ames, Iowa between 20016 and 2010.

More information about each variable can be found here:
http://jse.amstat.org/v19n3/decock/DataDocumentation.txt



1. Open on a new Colab notebook and import the pandas and altair packages

In [0]:
import pandas as pd
import altair as alt

2. Assign the link to the dataset to a variable called 'file_url'




In [0]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter12/Dataset/ames_iowa_housing.csv'

3. Using the read_csv .method() from the package pandas, load the dataset into a new DataFrame called 'df'

In [0]:
df = pd.read_csv(file_url)

4. Display the first 5 rows using the .head() method

In [4]:
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

5. Display the number of unique values on the column .nunique()

In [5]:
df['YearBuilt'].nunique()

112

There are 112 different values in the column ‘YearBuilt’.

6. Print a scatter plot using altair to visualise the number of records per year built. Specify 'YearBuilt:O' as x-axis and 'count()' as y-axis to the .encode() method

In [6]:
alt.Chart(df).mark_circle().encode(alt.X('YearBuilt:O'), y='count()')


alt.Chart(...)

There are not many properties sold for some of the years. We can group them by decades (group of 10 years).

7. Create a list called year_built containing all the unique values in the column 'YearBuilt'

In [0]:
year_built = df['YearBuilt'].unique()

8. Create another list that will compute the decade for each year in year_built. Use a list comprehension to loop through each year and apply the following formula: year - (year % 10). % corresponds to the modulo operator and will return the last digit of each year

In [0]:
decade_list = [year - (year % 10) for year in year_built]

9. Create a sorted list of unique values from decade_list and save the result into a new variable called decade_built. To do so, transform decade_list into a set (this will exclude all duplicates) and then used the sorted() function

In [0]:
decade_built = sorted(set(decade_list))

10. Print the values of decade_built

In [10]:
decade_built

[1870,
 1880,
 1890,
 1900,
 1910,
 1920,
 1930,
 1940,
 1950,
 1960,
 1970,
 1980,
 1990,
 2000,
 2010]

Now we have the list of decades we are going to bin the column ‘YearBuilt’ with.


11. Create a new column on the df DataFrame called 'DecadeBuilt' that will bin each value from 'YearBuilt' into a decade. You will use the .cut() method from pandas and specify the parameter bins=decade_built 

In [0]:
df['DecadeBuilt'] = pd.cut(df['YearBuilt'], bins=decade_built)

12. Print the first 5 rows of the DataFrame but only for the columns 'YearBuilt' and 'DecadeBuilt'

In [12]:
df[['YearBuilt', 'DecadeBuilt']].head()

YearBuilt   DecadeBuilt
0       2003  (2000, 2010]
1       1976  (1970, 1980]
2       2001  (2000, 2010]
3       1915  (1910, 1920]
4       2000  (1990, 2000]

Great Job! We can see that each year have been properly assigned to the relevant decade. We have successfully created a new feature from the column’YearBuilt’ by binning its values into group of decades. We have reduced the number of unique values from 112 to 15.